### PR TITLE
IGNITE-7101: .NET: IIgnite.GetVersion

### PR DESCRIPTION
**Pull request by ashapkin**
@ptupitsyn, please, take a look

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/PlatformProcessorImpl.java [hunk 1]
```diff
@@ -700,6 +704,17 @@ private void loggerLog(int level, String message, String category, String errorI
 
                 return;
             }
+
+            case OP_NODE_VERSION: {
+                IgniteProductVersion productVersion = ignite().cluster().node().version();
+                writer.writeByte(productVersion.major());
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/PlatformProcessorImpl.java]: Duplicates the code above. Please extract to `PlatformUtils.writeProductVersion` method.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/baseline/ClientAffinityAssignmentWithBaselineTest.java [hunk 2]
```diff
@@ -161,6 +163,8 @@
 
     /** {@inheritDoc} */
     @Override protected void beforeTestsStarted() throws Exception {
+        Assume.assumeFalse("https://issues.apache.org/jira/browse/IGNITE-11559", MvccFeatureChecker.forcedMvcc());
```
Comment (ptupitsyn) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/baseline/ClientAffinityAssignmentWithBaselineTest.java]: Looks like unrelated change to me.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Apache.Ignite.Core.Tests.csproj [hunk 3]
```diff
@@ -153,6 +153,7 @@
     <Compile Include="Client\RawSocketTest.cs" />
     <Compile Include="Client\ClientConnectionTest.cs" />
     <Compile Include="Client\IgniteClientConfigurationTest.cs" />
+    <Compile Include="Common\IgniteProductVersionTests.cs" />
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Apache.Ignite.Core.Tests.csproj]: Please add to `DotNetCore` test project as well.

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+        /** Revision hash. */
+        private readonly byte[] _revHash;
+
+        /// <summary>
+        /// Build a node version from <see cref="IBinaryRawReader"/>
+        /// </summary>
+        /// <param name="reader"><see cref="IBinaryRawReader"/></param>
+        /// <returns>Node version</returns>
+        public static IgniteProductVersion FromBinaryReader(IBinaryRawReader reader)
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Let's make this a constructor too for consistency with other classes.

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        /// </summary>
+        public byte Maintenance
+        {
+            get { return _maintenance; }
+        }
+
+        /// <summary>
+        /// Stage of development.
+        /// </summary>
+        public String Stage
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: `String` -> `string`

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            _major = major;
+            _minor = minor;
+            _maintenance = maintenance;
+            _stage = stage;
+            _revTs = revTs;
+            _revHash = revHash ?? new byte[20];
+        }
+        
+        /// <summary>
+        /// Major version number.
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Here and below: `Gets the ...`, see http://stylecop.soyuz5.com/SA1623.html

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+            return res;
+        }
+
+        /** <inheritDoc /> */
+        public override string ToString()
+        {
+            String revTsStr = ReleaseDate.ToString("yyyyMMdd");
+
+            return Major + "." + Minor + "." + Maintenance + "#" + revTsStr;
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Please use `string.Format`

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Common/IgniteProductVersionTests.cs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        [Test]
+        public void TestToString()
+        {
+            Assert.AreEqual("2.5.7#20180101", _defaultVersion.ToString());
+        }
+
+        [Test]
+        public void TestIgniteGetVersionMethod()
+        {
+            using (IIgnite ignite = Ignition.Start())
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Common/IgniteProductVersionTests.cs]: This does not work for multiple reasons (classpath, discovery settings, etc), please use `Ignition.Start(TestUtils.GetTestConfiguration())` in tests. Check what `TestUtils.GetTestConfiguration()` does.

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="major">Major version number.</param>
+        /// <param name="minor">Minor version number.</param>
+        /// <param name="maintenance">Maintenance version number.</param>
+        /// <param name="stage">Stage of development.</param>
+        /// <param name="revTs">Revision timestamp.</param>
+        /// <param name="revHash">Revision hash.</param>
+        public IgniteProductVersion(byte major, byte minor, byte maintenance, String stage, long revTs, byte[] revHash)
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Please don't use `String`, use `string` instead.

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+        /// </summary>
+        public string Stage
+        {
+            get { return _stage; }
+        }
+
+        /// <summary>
+        /// Gets the revision timestamp.
+        /// </summary>
+        public long RevisionTimestamp
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Let's remove this, `ReleaseDate` is enough

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        {
+            get { return _revTs; }
+        }
+
+        /// <summary>
+        /// Gets the release date.
+        /// </summary>
+        public DateTime ReleaseDate
+        {
+            get { return new DateTime(BinaryUtils.JavaDateTicks + _revTs * 1000); }
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Specify `DateTimeKind.Utc` to avoid ambiguity

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryUtils.cs [hunk 12]
```diff
@@ -69,7 +69,7 @@ internal static class BinaryUtils
         public const int ObjTypeId = -1;
 
         /** Ticks for Java epoch. */
-        private static readonly long JavaDateTicks = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc).Ticks;
+        public static readonly long JavaDateTicks = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc).Ticks;
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryUtils.cs]: Let's keep this encapsulated and create a method to convert Java timestamp to DateTime.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Common/IgniteProductVersionTests.cs [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+            IBinaryRawReader reader = SetUpRawBinaryReader(hash);
+
+            IgniteProductVersion deserializedVersion = new IgniteProductVersion(reader);
+            Assert.AreEqual("2.5.7#20180101-sha1:18", deserializedVersion.ToString());
+        }
+
+        [Test]
+        public void TestLocalNodeGetVersion()
+        {
+            using (IIgnite ignite = Ignition.Start(TestUtils.GetTestConfiguration()))
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Common/IgniteProductVersionTests.cs]: `Ignition.Start` is expensive. Please use `[TestFixtureSetUp]` to start a node once for all tests, or derive from `TestBase`.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Common/IgniteProductVersionTests.cs [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+        [Test]
+        public void TestClientVersionIsMoreOrEqualsServerNodeVersion()
+        {
+            using (IIgnite ignite = Ignition.Start(TestUtils.GetTestConfiguration()))
+            {
+                IgniteProductVersion version = ignite.GetVersion();
+
+                Version clientVer = Assembly.GetExecutingAssembly().GetName().Version;
+                var expectedVersion = new IgniteProductVersion((byte) clientVer.Major, (byte) clientVer.Minor,
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Common/IgniteProductVersionTests.cs]: Looks misleading.
Let's compare only major, minor, and patch for strict equality as three asserts.

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+        private readonly byte _maintenance;
+
+        /** Stage of development. */
+        private readonly String _stage;
+
+        /** Revision timestamp. */
+        private readonly long _revTs;
+
+        /** Revision hash. */
+        private byte[] _revHash;
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Should be `readonly`

File: modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+        /// </summary>
+        public byte Maintenance
+        {
+            get { return _maintenance; }
+        }
+
+        /// <summary>
+        /// Gets the stage of development.
+        /// </summary>
+        public string Stage
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Please add a test for this property.
It always returns `SNAPSHOT` for test builds.

Comment (ashapkin) [reply modules/platforms/dotnet/Apache.Ignite.Core/Common/IgniteProductVersion.cs]: Okay, but we already have `TestBinaryConstructor` method that uses `_defaultVersion `with populated Stage property under the hood.
